src/simulators/statevector/qubitvector_thrust.hpp

```diff
@@ -0,0 +1,3146 @@
+
+  // Returns a copy of the underlying data_t data class
+  std::complex<data_t>* data() const {return data_;}
+
+  //-----------------------------------------------------------------------
+  // Utility functions
+  //-----------------------------------------------------------------------
+
+  // Set the size of the vector in terms of qubit number
+  virtual void set_num_qubits(size_t num_qubits);
+
+  // Returns the number of qubits for the current vector
+  virtual uint_t num_qubits() const {return num_qubits_;}
+
+  // Returns the size of the underlying n-qubit vector
+  uint_t size() const {return data_size_;}
+
+  // Returns required memory
+  size_t required_memory_mb(uint_t num_qubits) const;
+
+  // Returns a copy of the underlying data_t data as a complex vector
+  cvector_t<data_t> vector() const;
+
+  // Return JSON serialization of QubitVectorThrust;
+  json_t json() const;
+
+  // Set all entries in the vector to 0.
+  void zero();
+
+  // convert vector type to data type of this qubit vector
+  cvector_t<data_t> convert(const cvector_t<double>& v) const;
+
+  // index0 returns the integer representation of a number of bits set
+  // to zero inserted into an arbitrary bit string.
+  // Eg: for qubits 0,2 in a state k = ba ( ba = 00 => k=0, etc).
+  // indexes0([1], k) -> int(b0a)
+  // indexes0([1,3], k) -> int(0b0a)
+  // Example: k = 77  = 1001101 , qubits_sorted = [1,4]
+  // ==> output = 297 = 100101001 (with 0's put into places 1 and 4).
+  template<typename list_t>
+  uint_t index0(const list_t &qubits_sorted, const uint_t k) const;
+
+  // Return a std::unique_ptr to an array of of 2^N in ints
+  // each int corresponds to an N qubit bitstring for M-N qubit bits in state k,
+  // and the specified N qubits in states [0, ..., 2^N - 1]
+  // qubits_sorted must be sorted lowest to highest. Eg. {0, 1}.
+  // qubits specifies the location of the qubits in the returned strings.
+  // NOTE: since the return is a unique_ptr it cannot be copied.
+  // indexes returns the array of all bit values for the specified qubits
+  // (Eg: for qubits 0,2 in a state k = ba:
+  // indexes([1], [1], k) = [int(b0a), int(b1a)],
+  // if it were two qubits inserted say at 1,3 it would be:
+  // indexes([1,3], [1,3], k) -> [int(0b0a), int(0b1a), int(1b0a), (1b1a)]
+  // If the qubits were passed in reverse order it would swap qubit position in the list:
+  // indexes([3,1], [1,3], k) -> [int(0b0a), int(1b0a), int(0b1a), (1b1a)]
+  // Example: k=77, qubits=qubits_sorted=[1,4] ==> output=[297,299,313,315]
+  // input: k = 77  = 1001101
+  // output[0]: 297 = 100101001 (with 0's put into places 1 and 4).
+  // output[1]: 299 = 100101011 (with 0 put into place 1, and 1 put into place 4).
+  // output[2]: 313 = 100111001 (with 1 put into place 1, and 0 put into place 4).
+  // output[3]: 313 = 100111011 (with 1's put into places 1 and 4).
+  indexes_t indexes(const reg_t &qubits, const reg_t &qubits_sorted, const uint_t k) const;
```

Comment:
IT doesn't look like `index0` or `indexes` functions are used anywhere in code, so could they be removed?

Reply:
removed